sympy/liealgebras/cartan_base.py

```diff
@@ -0,0 +1,295 @@
+            current.append(orb)
+            orbits_by_level[level] = current
+
+        sorbits = []
+        for level in sorted(orbits_by_level.keys()):
+            sorbits += sorted(orbits_by_level[level], key=lambda x: tuple(x))
+
+        # rotate back to the orthogonal basis for consistency
+        omega_matrix = self.omega_matrix()
+        return [x * omega_matrix for x in sorbits]
+
+    def roots(self):
+        """Returns the number of total roots in the algebra"""
+        raise NotImplementedError("Do not call this method directly from the base class.")
+
+    def root_level(self, root, basis='orthogonal'):
+        """Returns the root level of the root. The root level is calculated
+        by rotating by the omega matrix and then summing the rotated vector.
+        The basis argument flags which basis the incoming root or weight is in.
+        This class's default basis is 'orthogonal' but when this method is called
+        in the rootsystem calculations, the weights are passed via the 'alpha'
+        basis.
+
+        Examples
+        ========
+        >>> from sympy.liealgebras import CartanType
+        >>> g2 = CartanType("G2")
+        >>> rs = g2.rootsystem()
+        >>> g2.root_level(rs[0])
+        5
+        """
+
+        try:
+            self._cached_r
+        except AttributeError:
+            inverse_cartan = self.cartan_matrix().pinv()
+            self._cached_r = inverse_cartan * ones(inverse_cartan.rows, 1)
+
+        r = self._cached_r
+
+        if basis == 'orthogonal':
+            root = root * self.omega_matrix().pinv()
+
+        return (root * r)[0]
+
+    def positive_roots(self):
+        """Returns the set of all positive roots that are
+        """
+        n_pos = self.roots() // 2
+        return self.rootsystem()[:n_pos]
+
+    def orbit(self, head, stabilizer=None):
+        """
+        Returns the orbit of the weight or root by reflecting it
+        a plane. A stabilizer may be passed to calculate the orbit using
+        the Orbit-Stabilizer theorem.
+
+        Note the stabilizer starts counting simple roots at index 0
+
+        Sources
+        =======
+        - https://en.wikipedia.org/wiki/Coadjoint_representation#Coadjoint_orbit
+        - https://en.wikipedia.org/wiki/Group_action#Orbits_and_stabilizers
+        """
+        simple_roots = self.simple_roots()
+
+        if isinstance(head, MutableDenseMatrix):
+            head = head.as_immutable()
+
+        master_list = [head]
+        master_hash = set([tuple(head)])
+
+        if stabilizer:
+            reflect_loop = [simple_roots[i] for i in stabilizer]
+        else:
+            reflect_loop = simple_roots
+
+        # reflecting about a hyperplane
+        reflection_matrix = lambda v: eye(len(v)) - 2 * v.T * v / v.dot(v)
+
+        # Fill up master list of reflected roots by
+        # continually operating on them until all weights in orbit are found
+
+        # Duplicating each vector as a tuple to allow
```

Comment:
Is this necessary? `ImmutableDenseMatrix` can also be hashed. Could `master_hash` be initially `set((head,))`?

Reply:
I think this might have been a product of my inexperience with the library, let me try your suggestion.